app/components/avo/actions_component.rb

```diff
@@ -2,40 +2,47 @@
 
 class Avo::ActionsComponent < Avo::BaseComponent
   include Avo::ApplicationHelper
-  attr_reader :label, :size, :as_row_control
-
-  def initialize(actions: [], resource: nil, view: nil, exclude: [], include: [], style: :outline, color: :primary, label: nil, size: :md, as_row_control: false, icon: nil)
-    @actions = actions || []
-    @resource = resource
-    @view = view
-    @exclude = Array(exclude)
-    @include = include
-    @color = color
-    @style = style
-    @label = label || I18n.t("avo.actions")
-    @size = size
-    @icon = icon
-    @as_row_control = as_row_control
+
+  ACTION_FILTER = _Set(_Class(Avo::BaseAction))
+
+  prop :actions, _Array(Avo::BaseAction), default: -> { [] }
+  prop :resource, _Nilable(Avo::BaseResource)
+  prop :view, _Nilable(Symbol), &:to_sym
+  prop :exclude, _Nilable(ACTION_FILTER), &:to_set
+  prop :include, _Nilable(ACTION_FILTER), &:to_set
+  prop :style, Avo::ButtonComponent::STYLE, default: :outline
+  prop :color, Symbol, default: :primary
+  prop :label, String do |value|
+    value || I18n.t("avo.actions")
+  end
+  prop :size, Avo::ButtonComponent::SIZE, default: :md
+  prop :as_row_control, _Boolean, default: false
+  prop :icon, _Nilable(String)
+
+  def after_initialize
+    filter_actions
   end
 
   def render?
-    actions.present?
+    @actions.present?
   end
 
-  def actions
-    if @exclude.present?
-      @actions.reject { |action| action.class.in?(@exclude) }
-    elsif @include.present?
-      @actions.select { |action| action.class.in?(@include) }
-    else
-      @actions
+  def filter_actions
+    @actions = @actions.dup
+
+    if @exclude.any?
+      @actions.reject! { |action| @exclude.include?(action.class) }
+    end
+
+    if @include.any?
+      @actions.select! { |action| @include.include?(action.class) }
     end
   end
```

Comment:
I’ve changed action filtering to happen ahead of time. I’ve also updated the conditional so that you can filter with an include and exclude list. Previously, if there was an exclude filter, the include filters would be ignored.

The dup on line 31 is so we can use `reject!` and `select!` without concerns for mutating an array that might be used for something else after being passed to this component.

The filters uses `Set#include?` now, which will be really fast in the case that there are many actions to filter.

I’m not sure if any tests cover this functionality so may be one to look out for.

Reply:
This is great! Thanks!

Is it possible to access other props in one prop initializer?

Something like:
```ruby
  prop :actions, _Array(Avo::BaseAction), default: [].freeze do |actions|
    # Apply all the include / exclude logic here
    if @exclude.any?
      actions.reject! { |action| @exclude.include?(action.class) }
    end

    if @include.any?
      actions.select! { |action| @include.include?(action.class) }
    end

    actions
  end
```